# Sheetal Patil -  Estimations and Confidence Interval

**Background**
In quality control processes, especially when dealing with high-value items, destructive sampling is a necessary but costly method to ensure product quality. The test to determine whether an item meets the quality standards destroys the item, leading to the requirement of small sample sizes due to cost constraints.\
**Scenario**\
A manufacturer of print-heads for personal computers is interested in estimating the mean durability of their print-heads in terms of the number of characters printed before failure. To assess this, the manufacturer conducts a study on a small sample of print-heads due to the destructive nature of the testing process.\
**Data**\
A total of 15 print-heads were randomly selected and tested until failure. The durability of each print-head (in millions of characters) was recorded as follows:\
1.13, 1.55, 1.43, 0.92, 1.25, 1.36, 1.32, 0.85, 1.07, 1.48, 1.20, 1.33, 1.18, 1.22, 1.29


In [10]:
import numpy as np
import scipy.stats as stats

The command import **scipy.stats as stats** is a standard Python statement used to import the stats sub-package from the SciPy library and assign it the common alias stats.\
The scipy.stats module provides a wide range of functionality for statistical analysis, including:\
**Probability Distributions:** Access to over 80 continuous and 10 discrete random variable classes, such as the normal distribution (stats.norm) or binomial distribution (stats.binom). These objects have methods for:\
**pdf/pmf:** Probability density/mass function.\
**rvs:** Generating random variates (samples).\
**ppf:** Percent point function (inverse of CDF).\
**fit:** Estimating distribution parameters from data.\

In [11]:
# Data
data = np.array([1.13, 1.55, 1.43, 0.92, 1.25, 1.36, 1.32,
                 0.85, 1.07, 1.48, 1.20, 1.33, 1.18, 1.22, 1.29])

The most common way to create a numpy array is by converting a Python list or tuple using the np.array() function.

A **numpy array** is the core data structure in the NumPy library, a powerful tool for numerical computing in Python. Unlike standard Python lists, numpy arrays are designed for high-performance operations on large, multi-dimensional datasets.

**Key Characteristics**\
**Homogeneous:** All elements in a numpy array must be of the same data type (e.g., all integers or all floats), which allows for efficient storage and computation.\
**Multi-dimensional:** Arrays can have any number of dimensions (rank), making them suitable for representing vectors, matrices, and tensors.\
**Efficient Operations:** numpy operations are "vectorized," meaning they are applied to the entire array element-by-element without explicit Python loops, which is significantly faster than using standard Python lists.\
**Fixed Size:** Once created, the total size of an array cannot change.

In [12]:
# Sample size
n = len(data)

The **len() function** in Python is a built-in utility that returns the number of items in an object. It is a simple and efficient way to determine the size of various data structures and operates in constant time (O(1)) for most built-in types because the length is stored as an internal attribute.

In [13]:
# Sample statistics
mean = np.mean(data)
std_dev = np.std(data, ddof=1)   # sample standard deviation
alpha = 0.01

The **np.mean()** function in Python, part of the NumPy library, calculates the arithmetic mean (average) of elements in a given array-like object. It is a fundamental statistical tool used in data analysis, scientific computing, and machine learning.

The **np.std()** function in Python is used to compute the standard deviation of elements in a given array, which is a measure of the dispersion or spread of the data around its mean.\
Import NumPy as np. The general syntax is **numpy.std** (a, axis=None, dtype=None, out=None, ddof=0, keepdims=False). Key parameters include a (input array), axis (optional, specifies calculation axis, default is flattened), and ddof (optional, Delta Degrees of Freedom, default is 0).

In [20]:
print("Sample Mean:", round(mean, 4))
print("Sample Standard Deviation:", round(std_dev, 4))

Sample Mean: 1.2387
Sample Standard Deviation: 0.1932


In Python, the primary way to round numbers is the built-in **round() function**, which uses "banker's rounding" (round half to even).\
round(number, ndigits)\
number (required): The value to be rounded (integer or float).\
ndigits (optional): The number of decimal places to round to. The default is 0, which rounds to the nearest integer.

### (a) 99% CI using t-distribution

In [15]:
t_critical = stats.t.ppf(1 - alpha/2, df=n-1)
margin_t = t_critical * (std_dev / np.sqrt(n))
ci_t = (mean - margin_t, mean + margin_t)

The **scipy.stats.t.ppf()** function in Python calculates the percent point function (PPF), which is the inverse of the cumulative distribution function (CDF) for the Student's t-distribution. It is used to find the critical value (t-score) corresponding to a given left-tail probability and degrees of freedom.\
**scipy.stats.t.ppf(q, df, loc=0, scale=1)**\
q: The lower tail probability or critical level (a value between 0 and 1).\
df: The degrees of freedom (e.g., sample size minus 1).\
loc: The location parameter (mean), which defaults to 0.\
scale: The scale parameter (standard deviation), which defaults to 1.

In [21]:
print("\n99% CI using t-distribution:", 
      (round(ci_t[0], 4), round(ci_t[1], 4)))


99% CI using t-distribution: (np.float64(1.0902), np.float64(1.3871))


**Interpretation:**\
We are 99% confident that the true mean durability of the print-heads lies between 1.090 and 1.387 million characters.

### (b) 99% CI using z-distribution

In [16]:
z_critical = stats.norm.ppf(1 - alpha/2)
margin_z = z_critical * (std_dev / np.sqrt(n))  # assuming sigma = sample std
ci_z = (mean - margin_z, mean + margin_z)

The **scipy.stats.norm.ppf()** function in Python computes the percent-point function (PPF), which is the inverse of the cumulative distribution function (CDF) for a normal distribution. It is used to find the value of a random variable x that corresponds to a given cumulative probability (quantile) q.\
**scipy.stats.norm.ppf(q, loc=0, scale=1)**\
Parameters\
q: The cumulative probability (quantile) for which to compute the value of x. This can be a single float or a NumPy array of probabilities.\
loc (optional): The mean () of the normal distribution. The default is 0 (standard normal distribution).\
scale (optional): The standard deviation () of the normal distribution. The default is 1 (standard normal distribution).

In [22]:
print("99% CI using z-distribution:", 
      (round(ci_z[0], 4), round(ci_z[1], 4)))

99% CI using z-distribution: (np.float64(1.1102), np.float64(1.3671))


**Interpretation:**
With known population variability, we are 99% confident that the true mean durability lies between 1.106 and 1.372 million characters.